### PR TITLE
Hydro Steam

### PR DESCRIPTION
**Pull request by Fontbane**
Split #159 for now. This part includes the Hydro Steam implementation, which is a little different since some things have changed. Also includes Spirit Shackle, and a fix for a bug(?) I noticed where a wild mon using U-turn would result in it fleeing.

### CODE REVIEW

File: src/data/move.ts [hunk 1]
```diff
@@ -2870,6 +2886,9 @@ export class ForceSwitchOutAttr extends MoveEffectAttr {
       if (!player && !user.scene.currentBattle.battleType) {
         if (this.batonPass)
           return false;
+        //U-turn et al should not switch a wild mon out, but a player's Dragon Tail can
+        if (!(user instanceof PlayerPokemon) && move.category!=MoveCategory.STATUS)
```
Comment (Flashfyre) [inline src/data/move.ts:2890]:
```suggestion
        if (!(user instanceof PlayerPokemon) && move.category !== MoveCategory.STATUS)
```
Minor formatting nitpick

**Review by Flashfyre — CHANGES_REQUESTED** [review]

File: src/data/move.ts (hunk 1 above)
Comment (Xavion3) [reply src/data/move.ts]: This should probably use hasTrainer too right?

**Comment by EmoUsedHM01** [comment]
The implementation of Hydro Steam is incorrect
Firstly the fix has already been PR'd (#123)
Secondly it should get a damage boost as if it was a fire type move under sun

File: src/data/move.ts [hunk 2]
```diff
@@ -789,6 +789,22 @@ export class SacrificialFullRestoreAttr extends SacrificialAttr {
   }
 }
 
+export class IgnoreWeatherTypeDebuffAttr extends MoveAttr {
+  public weather: WeatherType;
+  constructor(weather: WeatherType){
+    super();
+    this.weather=weather;
```
Comment (Xavion3) [inline src/data/move.ts:796]: Should have spaces around the equals, just a small formatting thing.

File: src/data/move.ts [hunk 3]
```diff
@@ -789,6 +789,22 @@ export class SacrificialFullRestoreAttr extends SacrificialAttr {
...
+  }
+  apply(user: Pokemon, target: Pokemon, move: Move, args: any[]): boolean {
+    (args[0] as Utils.IntegerHolder).value = Math.max((args[0] as Utils.IntegerHolder).value, 1);
+
+    return true;
+  }
+  getCondition(): MoveCondition | MoveConditionFunc {
+      return (user, target, move) => user.scene.arena.weather?.weatherType==this.weather;
```
Comment (Xavion3) [inline src/data/move.ts:804]: Doesn't this cause the entire move to fail if there's no sun active?

File: src/data/move.ts [hunk 4]
```diff
@@ -2870,6 +2886,9 @@ export class ForceSwitchOutAttr extends MoveEffectAttr {
       if (!player && !user.scene.currentBattle.battleType) {
         if (this.batonPass)
           return false;
+        //U-turn et al should not switch a wild mon out, but a player's Dragon Tail can
+        if (!(user instanceof PlayerPokemon) && move.category !== MoveCategory.STATUS)
```
Comment (Xavion3) [inline src/data/move.ts:2890]: It should still work for enemy trainers, there should be a hasTrainer function.

**Review by Xavion3 — CHANGES_REQUESTED** [review]

**Comment by Fontbane** [comment]
Should work properly now.

I also noticed there's more to forced switch out moves than we have implemented, including different interactions in double wild battles and taking levels into account. Might open an issue for that.

**Comment by Fontbane** [comment]
Okay, Parting Shot is more involved than I thought. It only switches out if the target gets its stats lowered, or if the stat changes get bounced back via Mirror Armor. If used on a target with Magic Bounce, it lowers the user's stats and switches out the target. It also works even if it the user has nothing to switch to, in which case it just works as a regular stat lowering move.

Chilly Reception is similar, but different in that neither of the effects are dependent on each other, so it will still switch out even if snow can't be set up, and it will set up snow even if it can't switch out, but if it can't do either, the move fails... _but_ before it fails it will still display the message, "(user) is preparing to tell a chillingly bad joke!".

Teleport is the only self switch out move which allows the user to flee if it's a wild Mon.

**Comment by bennybroseph** [comment]
Sorry, is that explanation for how it **does** work in the code you submitted or how it **should** work in the mainline series and more changes are still required?

**Review by bennybroseph — APPROVED** [review]

**Review by bennybroseph — APPROVED** [review]

**Review by Madmadness65 — APPROVED** [review]

**Review by Xavion3 — APPROVED** [review]

**Review by Flashfyre — APPROVED** [review]

**Pull request merged**